feat(cache): consolidate caching API and strategies

hypeJunction opened the pull request.

hypeJunction commented:
I can't see any more Elgg code. This is good for review, need to figure what's up with tests. I am leaving it all for a while, I need a break

mrclay commented on engine/classes/Elgg/Database/EntityTable.php:
```diff
@@ -627,4 +626,110 @@ public function disableEntities($owner_guid) {
+
+		_elgg_services()->hooks->getEvents()->triggerAfter('delete', $entity->type, $entity);
+
+		return (bool) $deleted;
```
When would we want to return false?

hypeJunction replied:
I am not sure. Just moved this from ElggEntity

mrclay commented on engine/classes/Elgg/Cache/MetadataCache.php:
```diff
@@ -227,39 +254,16 @@ public function populateFromEntities(...$guids) {
-				$this->ids[$guid][$name] = $id;
-			}
+			$row->value = ($row->value_type === 'text') ? $row->value : (int) $row->value;
+			$metadata = new ElggMetadata($row);
```
What do we get for the overhead of creating these objects?

hypeJunction replied:
Not sure how we would measure it. Definitely better than iterating through bunch of array and querying the DB to compare metadata names and ids

hypeJunction replied:
We can also just cache attributes

mrclay commented on engine/classes/Elgg/Database/PrivateSettingsTable.php:
```diff
@@ -132,29 +134,40 @@ public function getAll($entity_guid) {
-			':name' => (string) $name,
-			':value' => (string) $value,
-		];
+		_elgg_invalidate_cache_for_entity($entity_guid);
```
This stays?

mrclay replied:
That's fine

mrclay reviewed (commented):
Bad idea for me to try reviewing this on my tiny phone!

hypeJunction commented on engine/classes/Elgg/Cache/MetadataCache.php:
```diff
@@ -270,25 +274,25 @@ public function populateFromEntities(...$guids) {
 			'order_by' => 'n_table.entity_guid, n_table.time_created ASC',
 			'group_by' => 'n_table.entity_guid',
+			'having' => [
+				"bytes > $limit OR bytes < 0",
```
Should be the opposite. Need to filter those that have heavy meta

hypeJunction commented on engine/classes/Elgg/Cache/SessionCache.php:
```diff
@@ -0,0 +1,21 @@
+	 * @return CompositeCache
+	 */
+	protected function create($namespace) {
+		return new CompositeCache($namespace, $this->config, ELGG_CACHE_RUNTIME);
```
I am contemplating making it persistent

hypeJunction commented on engine/classes/Elgg/Database/EntityTable.php:
```diff
@@ -627,4 +626,110 @@ public function disableEntities($owner_guid) {
+		$ia = elgg_set_ignore_access(true);
+
+		// Now delete the entity itself
+		$entity->deleteMetadata();
```
Ned to add foreign key and get rid of this madness

hypeJunction commented on engine/classes/Elgg/Database/PrivateSettingsTable.php:
```diff
@@ -99,21 +96,24 @@ public function get($entity_guid, $name) {
-			':entity_guid' => (int) $entity_guid,
-		];
+		$values = $this->cache->load($entity_guid);
+		if ($values) {
```
Eh?

hypeJunction commented on engine/classes/Elgg/Database/PrivateSettingsTable.php:
```diff
@@ -132,29 +134,40 @@ public function getAll($entity_guid) {
+			->andWhere($qb->compare('entity_guid', '=', $entity_guid, ELGG_VALUE_INTEGER));
 
-		$result = $this->db->insertData($query, $params);
+		$row = $this->db->getDataRow($qb);
```
Not really efficient. Better strategy would be to cache settings with as database rows and grab IDs from there

hypeJunction commented on engine/classes/ElggUpgrade.php:
```diff
@@ -15,6 +15,11 @@
  * @package Elgg.Admin
  * @access private
+ *
+ * @property int $is_complete
```
Completed

hypeJunction commented:
I give up. All tests are passing locally, but not on Travis. Faster tests come with all that shitty global state that spills over.

hypeJunction commented:
Tests are green now

jdalsem commented on 1:
```diff
@@ -0,0 +1,21 @@
+pick 2e5fe21b8 feat(cache): consolidate caching API and strategies
```
where is this file coming from

hypeJunction replied:
Not sure. Something happened during rebase. Will remove

hypeJunction commented on engine/classes/Elgg/Database/EntityTable.php:
```diff
@@ -455,9 +464,16 @@ public function fetch(QueryBuilder $query, array $options = []) {
 		});
+		/* @var $preload ElggEntity[] */
+
+		if ((bool) $options['limit'] === false || $options['limit'] > elgg_get_config('default_limit')) {
```
Allows allow plugins

hypeJunction commented on engine/classes/ElggPlugin.php:
```diff
@@ -1155,8 +1149,9 @@ private function setStatus($active) {
 		}
 
+		$this->invalidateCache();
+		_elgg_services()->plugins->invalidateCache($this->getID());
```
Move to entity method

hypeJunction reviewed (commented):
Need to bring back entity cache. Locking session cache creates a situation where caches are not invalidated when entity is updated with ignored access

hypeJunction committed:
feat(cache): consolidate caching API

Caching has been consolidated using Stash pools.
Invalidation strategies have been reviewed and centralized
around major entity events.

hypeJunction commented:
I think I got it all sorted out now.

mrclay commented:
Works fine for me.

jeabakker commented:
This PR accidentally (I assume) undid #11503
